core/txpool/legacypool/list.go

```diff
@@ -188,15 +195,16 @@ func (m *SortedMap) Remove(nonce uint64) bool {
 		return false
 	}
 	// Otherwise delete the transaction and fix the heap index
-	for i := 0; i < m.index.Len(); i++ {
-		if (*m.index)[i] == nonce {
-			heap.Remove(m.index, i)
-			break
-		}
-	}
+	index := slices.Index(*m.index, nonce)
+	heap.Remove(m.index, index)
 	delete(m.items, nonce)
+
 	m.cacheMu.Lock()
-	m.cache = nil
+	if m.cache != nil {
+		slices.DeleteFunc(m.cache, func(tx *types.Transaction) bool {
```

Comment:
```go
// DeleteFunc removes any elements from s for which del returns true,
// returning the modified slice.
// DeleteFunc zeroes the elements between the new length and the original length.
func DeleteFunc[S ~[]E, E any](s S, del func(E) bool) S {
	i := IndexFunc(s, del)
	if i == -1 {
		return s
	}
	// Don't start copying elements until we find one to delete.
	for j := i + 1; j < len(s); j++ {
		if v := s[j]; !del(v) {
			s[i] = v
			i++
		}
	}
	clear(s[i:]) // zero/nil out the obsolete elements, for GC
	return s[:i]
}
```

the Del function will be applied on the all following elements, which is also not cheap.